packages/file-search/src/node/file-search-service-impl.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2017 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { FileSearchService } from '../common/file-search-service';
+import * as fs from "fs";
+import * as path from "path";
+import URI from '@theia/core/lib/common/uri';
+import { FileUri } from '@theia/core/lib/node';
+import * as fuzzy from 'fuzzy';
+import { injectable, inject } from 'inversify';
+import * as parser from 'gitignore-parser';
+import { ILogger } from '@theia/core';
+
+type Filtered = (simpleName: string, fullPath: string) => boolean;
+
+@injectable()
+export class FileSearchServiceImpl implements FileSearchService {
+
+    constructor( @inject(ILogger) private logger: ILogger) { }
+
+    async find(uri: string, searchPattern: string, options?: { fuzzyMatch?: boolean, limit?: number, useGitignore?: boolean }): Promise<string[]> {
+        const basePath = FileUri.fsPath(new URI(uri));
+        const opts = {
+            fuzzyMatch: true,
+            limit: Number.MAX_SAFE_INTEGER,
+            useGitignore: true,
+            ...options
+        };
+        const result: string[] = [];
+        const limitReached = new Error("limit reached");
+        const filtered = this.getFiltered(basePath, opts);
+        try {
+            this.findRecursive(basePath, filtered, filePath => {
+                if (result.length >= opts.limit) {
+                    throw limitReached;
+                }
+                if (opts.fuzzyMatch && fuzzy.test(searchPattern, filePath)) {
+                    result.push(FileUri.create(filePath).toString());
+                } else {
+                    if (filePath.toLocaleLowerCase().indexOf(searchPattern.toLocaleLowerCase()) !== -1) {
+                        result.push(FileUri.create(filePath).toString());
+                    }
+                }
+            });
+        } catch (e) {
+            if (e !== limitReached) {
+                throw e;
```

Comment:
This makes it so that let's say you do not have perms to read a file in a sub-directory the whole search will fail.

We should handle errors by fs.readDirSync and treat them as warnings. Or just swallow them.

Reply:
ok